patches/0013-p256-armv8-optimizations

```diff
@@ -2429,20 +2457,446 @@ Index: aws-lc/third_party/boringssl/crypto/fipsmodule/ec/p256-x86_64_test.cc
++#                                     BN_ULONG n[P256_LIMBS]);
++#
++# (Binary Extended GCD (Euclidean) Algorithm.
++#  See A. Menezes, P. vanOorschot, and S. Vanstone's Handbook of Applied Cryptography,
++#  Chapter 14, Algorithm 14.61 and Note 14.64
++#  http://cacr.uwaterloo.ca/hac/about/chap14.pdf)
++#
++# Assumption 1: n is odd for the BEEU
++# Assumption 2: 1 < a < n < 2^256
++
++# Same algorithm in Python:
++# def beeu(a, n):
++#     X = 1
++#     Y = 0
++#     B = a
++#     A = n
++#     while (B != 0):
++#         while (B % 2) == 0:
++#             B >>= 1
++#             if (X % 2) == 1:
++#                 X = X + n
++#             X >>= 1
++#         while (A % 2) == 0:
++#             A >>= 1
++#             if (Y % 2) == 1:
++#                 Y = Y + n
++#             Y >>= 1
++#         if (B >= A):
++#             B = B - A
++#             X = X + Y
++#         else:
++#             A = A - B
++#             Y = Y + X
++#     if (A != 1):
++#         # error
++#         return 0
++#     else:
++#         while (Y > n):
++#             Y = Y - n
++#         Y = n - Y
++#         return Y
++
++# For the internal variables,
++# x0-x2, x30 are used to hold the modulus n. The input parameters passed in
++# x1,x2 are copied first before corrupting them. x0 (out) is stored on the stack.
++# x3-x7 are used for parameters, which is not the case in this function, so they are corruptible
++# x8 is corruptible here
++# (the function doesn't return a struct, hence x8 doesn't contain a passed-in address
++#  for that struct).
++# x9-x15 are corruptible registers
++# x19-x28 are callee-saved registers
++
++# X/Y will hold the inverse parameter
++# Assumption: a,n,X,Y < 2^(256)
++# Initially, X := 1, Y := 0
++#            A := n, B := a
++
++# Function parameters (as per the Procedure Call Standard)
++my($out, $a_in, $n_in)=map("x$_",(0..2));
++# Internal variables
++my($n0, $n1, $n2, $n3)=map("x$_",(0..2,30));
++my($x0, $x1, $x2, $x3, $x4)=map("x$_",(3..7));
++my($y0, $y1, $y2, $y3, $y4)=map("x$_",(8..12));
++my($shift)=("x13");
++my($t0, $t1, $t2, $t3)=map("x$_",(14,15,19,20));
++my($a0, $a1, $a2, $a3)=map("x$_",(21..24));
++my($b0, $b1, $b2, $b3)=map("x$_",(25..28));
++
++# if B == 0, jump to end of loop
++sub TEST_B_ZERO {
++  return <<___;
++    orr     $t0, $b0, $b1
++    orr     $t0, $t0, $b2
++
++    // reverse the bit order of $b0. This is needed for clz after this macro
++    rbit     $t1, $b0
++
++    orr     $t0, $t0, $b3
++    cbz     $t0,.Lbeeu_loop_end
++___
++}
++
++# Shift right by 1 bit, adding the modulus first if the variable is odd
++# if least_sig_bit(var0) == 0,
++#     goto shift1_<ctr>
++# else
++#     add n and goto shift1_<ctr>
++# Prerequisite: t0 = 0
++$g_next_label = 0;
++sub SHIFT1 {
++  my ($var0, $var1, $var2, $var3, $var4) = @_;
++  my $label = ".Lshift1_${g_next_label}";
++  $g_next_label++;
++  return <<___;
++    tbz     $var0, #0, $label
++    adds    $var0, $var0, $n0
```

Comment:
In general, you wouldn't want to add the carry in the least significant word unless you've explicitly set it to 0. Instead of doing that, you can simply ignore it.